### DISCUSSION TITLE
build error

### DISCUSSION THREAD

**Discussion by dewan1988** [idaholab/moose#26676] [Q&A Modules: General]
Dear MOOSE experts
I am trying to install moose for HPC. at first I created conda moose environment for LInux users by following commands of the link: https://mooseframework.inl.gov/getting_started/installation/conda.html then I followed the instructions are given for those operating on an HPC cluster: https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html. but when I am trying to update and rebuild the libmesh.sh highlited here
![image](https://github.com/idaholab/moose/assets/145499408/035e3288-387f-44ce-9695-7b69200a09db)
it is showing following errors:

```
PETSc submodule will be used. PETSc submodule is our default solver.
IMPORTANT: If you did not run the update_and_rebuild_petsc.sh script yet, please run it before building libMesh
---------------------------------------------
----------- Configuring libMesh -------------
---------------------------------------------
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /home/mdewan/projects/moose/scripts/../libmesh/build-aux/install-sh -C
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether UID '1108' is supported by ustar format... yes
checking whether GID '1047' is supported by ustar format... yes
checking how to create a ustar tar archive... gnutar
checking whether make supports nested variables... (cached) yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for src/base/libmesh.C... no
<<< Configuring build directory for VPATH build >>>
checking for perl... /usr/bin/perl
note: MPI library path not given...
trying prefix=/usr/lib64/compat-openmpi16
note: MPI library path not given...
trying prefix=/usr/lib64/compat-openmpi16
checking whether make supports the include directive... yes (GNU style)
checking whether the C compiler works... no
configure: error: in `/home/mdewan/projects/moose/libmesh/build':
configure: error: C compiler cannot create executables
See `config.log' for more details
Running make -j 6...
make: *** No targets specified and no makefile found.  Stop.
```
Could you please help me?

**Comment by GiudGiud** [comment]
Hello

Please run the diagnostics script in moose/scripts
Seems the compiler being found is not working:
```
checking whether the C compiler works... no
```

Guillaume

**Reply by dewan1988** [reply]
Dear GiudGiud
It is showing following information:
System Arch: LSB Version: :core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch Distributor ID: CentOS Description: CentOS Linux release 7.9.2009 (Core) Release: 7.9.2009 Codename: Core

MOOSE Package Version: Custom Build

CPU Count: 16

Memory Free: 2250.941 MB

$CC not set

MPICC WRAPPER NOT FOUND

Python:
        /home/mdewan/miniforge/bin/python
        Python 3.10.13

MODULES:

Currently Loaded Modules:
  1) slurm/slurm/19.05.8   2) GCCcore/10.3.0   3) zlib/1.2.11-GCCcore-10.3.0   4) binutils/2.36.1-GCCcore-10.3.0   5) GCC/10.3.0   6) mpi/compat-openmpi16-x86_64

**Reply by GiudGiud** [reply]
You need to find mpicc, mpif90 and mpicxx on your system.
Once you have found those (likely coming from 6) mpi/compat-openmpi16-x86_64) you need to set:
```
export PATH=<path to mpicc etc>:${PATH}
export CC=mpicc
export CXX=mpicxx
export F90=mpif90
```

**Reply by milljm** [reply]
You'll also most likely need to rebuild PETSc as a result. It seems PETSc may have found some other compiler (I am assuming it finished since we're talking about `update_and_rebuild_libmesh.sh`

**Reply by dewan1988** [reply]
Dear GiudGiud
I got the location of mpicc, mpif90 and mpicxx on my system.
/home/mdewan/miniforge/envs/moose/bin 
But I didn't understand this line "Once you have found those (likely coming from 6) mpi/compat-openmpi16-x86_64) you need to set:"
because I didn't find " mpi/compat-openmpi16-x86_64" in this location. How can I set 
export PATH=<path to mpicc etc>:${PATH}
export CC=mpicc
export CXX=mpicxx
export F90=mpif90

**Reply by GiudGiud** [reply]
these are environment variables. You set or modify them using `export`

So you want to use miniforge to build? You have a mpi module (openmpi) loaded. you have to choose. On a HPC infrastructure, you should likely use the MPI module

**Reply by dewan1988** [reply]
Yes, this miniforge was automatically built up during conda environment creation
![image](https://github.com/idaholab/moose/assets/145499408/55962530-1d14-45d1-881a-dffe953e13a1)

**Reply by GiudGiud** [reply]
this is fine if you want to install this way but you cannot do that:

>  at first I created conda moose environment for LInux users by following commands of the link: https://mooseframework.inl.gov/getting_started/installation/conda.html then I followed the instructions are given for those operating on an HPC cluster: https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html. 

you have to pick one way and stick to it

**Reply by dewan1988** [reply]
I have set the following environment and tried to rebuild the  PETSc, libMesh, and WASP.
```
export PATH=<path to mpicc etc>:${PATH}
export CC=mpicc
export CXX=mpicxx
export F90=mpif90
```
![image](https://github.com/idaholab/moose/assets/145499408/8fb61373-722c-431b-8434-b4b443a46397)


Still it is showing the following error: can not find -ludev etc.

```
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libX11.so while searching for X11
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: error: cannot find -ludev
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libpthread.so while searching for pthread
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/librt.so while searching for rt
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libm.so while searching for m
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libdl.so while searching for dl
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libc.so while searching for c
collect2: error: ld returned 1 exit status
make[4]: *** [libnetcdf.la] Error 1
make[4]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib/netcdf/netcdf-c-4.6.2/liblib'
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib/netcdf/netcdf-c-4.6.2'
make[2]: *** [all] Error 2
make[2]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib/netcdf/netcdf-c-4.6.2'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib'
make: *** [all-recursive] Error 1
```

**Reply by GiudGiud** [reply]
you need to fill with the path to your compilers inside `< ... >` 
```
export PATH=<...>:${PATH}
```

**Reply by dewan1988** [reply]
I have deleted my previous installation and now trying to install the moose only focusing the instruction provided for the HPC : https://mooseframework.inl.gov/getting_started/installation/hpc_install_moose.html
But still I am stucking in the same place (rebuilding_libmesh) and getting the following errors:

![image](https://github.com/idaholab/moose/assets/145499408/a22c4468-8431-41c2-b39e-cc904548ff94)

```
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libX11.so while searching for X11
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: error: cannot find -ludev
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libpthread.so while searching for pthread
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/librt.so while searching for rt
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libm.so while searching for m
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libdl.so while searching for dl
/cm/shared/software/binutils/2.36.1-GCCcore-10.3.0/bin/ld.gold: warning: skipping incompatible /lib/libc.so while searching for c
collect2: error: ld returned 1 exit status
make[4]: *** [libnetcdf.la] Error 1
make[4]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib/netcdf/netcdf-c-4.6.2/liblib'
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib/netcdf/netcdf-c-4.6.2'
make[2]: *** [all] Error 2
make[2]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib/netcdf/netcdf-c-4.6.2'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/home/mdewan/projects/moose/libmesh/build/contrib'
make: *** [all-recursive] Error 1

```

**Reply by GiudGiud** [reply]
hello

seems you are missing some dependencies
can you run this?
```
sudo apt-get install libudev1 libudev-dev
```

or find a module on your HPC that provides it

**Reply by dewan1988** [reply]
But there is a warning about the sudo in the Conda MOOSE Environment creation. Is this for the Linux only? will not have any impact for HPC installation? 

![image](https://github.com/idaholab/moose/assets/145499408/0170ff14-e09a-4f44-adff-0c228f640040)

**Reply by GiudGiud** [reply]
you are not using the conda install path anymore. 
I think because you are on HPC you wont be able to use sudo and you will have to ask for the HPC team to help you install this

**Reply by milljm** [reply]
Don't mean to derail the thread... PETSc is usually the more finicky library, due to the myriad amount of contribs we are asking it to build. By any change can you attach the configure log for the PETSc build? It might help us better understand your cluster. The log file should reside at:

```pre
/home/mdewan/projects/moose/petsc/configure.log
```

**Reply by dewan1988** [reply]
There are lots of error. I just copied the last part:
 Loop over declared dependencies of package and error if any are missing
```
                    Defined make rule "slepcbuild" with dependencies "" and code ['@echo "*** Building SLEPc ***"', '@${RM} /home/mdewan/projects/moose/petsc/arch-moose/lib/petsc/conf/slepc.errorflg', '+@(cd /home/mdewan/projects/moose/petsc/arch-moose/externalpackages/git.slepc && \\\n            SLEPC_DIR=/home/mdewan/projects/moose/petsc/arch-moose/externalpackages/git.slepc /home/mdewan/miniforge/bin/python3 ./configure --prefix=/home/mdewan/projects/moose/petsc/arch-moose  && \\\n            SLEPC_DIR=/home/mdewan/projects/moose/petsc/arch-moose/externalpackages/git.slepc ${OMAKE} SLEPC_DIR=/home/mdewan/projects/moose/petsc/arch-moose/externalpackages/git.slepc) || \\\n            (echo "**************************ERROR*************************************" && \\\n            echo "Error building SLEPc." && \\\n            echo "********************************************************************" && \\\n            touch /home/mdewan/projects/moose/petsc/arch-moose/lib/petsc/conf/slepc.errorflg && \\\n            exit 1)']
/tmp/petsc-w1_gvvnk/PETSc.Configure/conftest.c:4:15: error: ‘RTLD_DEFAULT’ undeclared here (not in a function)
/tmp/petsc-w1_gvvnk/PETSc.Configure/conftest.c:4:15: error: ‘RTLD_DEFAULT’ undeclared here (not in a function)
  -pass-exit-codes         Exit with highest error code from a phase.
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/config.libraries/conftest.c:4:10: fatal error: windows.h: No such file or directory
/* Override any gcc2 internal prototype to avoid an error. */
/tmp/petsc-w1_gvvnk/PETSc.Configure/conftest.c:3:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/PETSc.Configure/conftest.c:3:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/PETSc.Configure/conftest.c:3:10: fatal error: windows.h: No such file or directory
/tmp/petsc-w1_gvvnk/PETSc.Configure/conftest.c:3:10: fatal error: windows.h: No such file or directory
  -pass-exit-codes         Exit with highest error code from a phase.
      Warning accessing  gives errors: [Errno 2] No such file or directory: ''
      Warning accessing  gives errors: [Errno 2] No such file or directory: ''
    If --with-strict-petscerrorcode = 1:
        @${RM} /home/mdewan/projects/moose/petsc/arch-moose/lib/petsc/conf/slepc.errorflg
            touch /home/mdewan/projects/moose/petsc/arch-moose/lib/petsc/conf/slepc.errorflg && \
```

**Reply by milljm** [reply]
There will naturally be lots of error looking lines. But what we really need is the attached log file. It can be several megabytes in size, so please do not even try to paste the text :).

If there actually was a failure in PETSc, then we need to stop there, and figure that out first, as libMesh _will_ fail to build if PETSc isn't complete.

**Reply by dewan1988** [reply]
During PETSc rebuilding I didn't get error, that is why I have tried to update and rebuilding the libmesh.

**Reply by milljm** [reply]
Considering what has been posted, we are concerned at the current state of the build. That is why we wish to see the PETSc configure log. PETSc does a real good job at hunting for things all the contribs need. But sometimes it finds things it shouldn't. By attaching the PETSc configure log, we can better understand that environment, and possibly why libMesh is failing to build.

Can you attach your PETSc configure log? It should be located at:

```pre
home/mdewan/projects/moose/petsc/configure.log
```
